src/os/bluestore/BlueFS.cc

```diff
@@ -1927,10 +1927,24 @@ int BlueFS::_allocate(uint8_t id, uint64_t len,
 
   uint64_t left = ROUND_UP_TO(len, min_alloc_size);
   int r = -ENOSPC;
+  int64_t alloc_len = 0;
+  AllocExtentVector extents;
+  
   if (alloc[id]) {
     r = alloc[id]->reserve(left);
   }
-  if (r < 0) {
+  
+  if (r == 0) {
+    uint64_t hint = 0;
+    if (!node->extents.empty() && node->extents.back().bdev == id) {
+      hint = node->extents.back().end();
+    }   
+    extents.reserve(4);  // 4 should be (more than) enough for most allocations
+    alloc_len = alloc[id]->allocate(left, min_alloc_size, hint, &extents);
+  }
+  if (r < 0 || (alloc_len < (int64_t)left)) {
+    if (r == 0)
+      alloc[id]->unreserve(left);
```

Comment:
You should probably unreserve the difference between left and alloc_len

Reply:
Ahh, right.


> release already allocated extents

But instead of doing this, I think we can just push the allocated extents into **ev** as we technically support mix-usage of space from different devices:
```c++
    if (r ==0) {
      alloc[id]->unreserve(left - alloc_len);
      for (auto& p : extents) {
        bluefs_extent_t e = bluefs_extent_t(id, p.offset, p.length);
        if (!ev->empty() &&
	    ev->back().bdev == e.bdev &&
	    ev->back().end() == (uint64_t) e.offset) {
          ev->back().length += e.length;
        } else {
          ev->push_back(e);
        }
      }
```

Reply:
Well, on a second thought, "release already allocated extents" perhaps is a better choice as supporting a single request of allocating space from multiple devices is a little bit too complicated...